nexus/src/app/sagas/instance_delete.rs

```diff
@@ -0,0 +1,106 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+use super::ActionRegistry;
+use super::NexusActionContext;
+use super::NexusSaga;
+use crate::app::sagas::declare_saga_actions;
+use crate::context::OpContext;
+use crate::{authn, authz};
+use serde::Deserialize;
+use serde::Serialize;
+use steno::ActionError;
+
+// instance delete saga: input parameters
+
+#[derive(Debug, Deserialize, Serialize)]
+pub struct Params {
+    pub serialized_authn: authn::saga::Serialized,
+    pub authz_instance: authz::Instance,
+}
+
+// instance delete saga: actions
+
+declare_saga_actions! {
+    instance_delete;
+    INSTANCE_DELETE_RECORD -> "no_result1" {
+        + sid_delete_instance_record
+    }
+    DELETE_NETWORK_INTERFACES -> "no_result2" {
+        + sid_delete_network_interfaces
+    }
+    DEALLOCATE_EXTERNAL_IP -> "no_result3" {
+        + sid_deallocate_external_ip
+    }
+}
+
+// instance delete saga: definition
+
+#[derive(Debug)]
+pub struct SagaInstanceDelete;
+impl NexusSaga for SagaInstanceDelete {
+    const NAME: &'static str = "instance-delete";
+    type Params = Params;
+
+    fn register_actions(registry: &mut ActionRegistry) {
+        instance_delete_register_actions(registry);
+    }
+
+    fn make_saga_dag(
+        _params: &Self::Params,
+        mut builder: steno::DagBuilder,
+    ) -> Result<steno::Dag, super::SagaInitError> {
+        builder.append(instance_delete_record_action());
+        builder.append(delete_network_interfaces_action());
+        builder.append(deallocate_external_ip_action());
+        Ok(builder.build()?)
+    }
+}
+
+// instance delete saga: action implementations
+
+async fn sid_delete_instance_record(
+    sagactx: NexusActionContext,
+) -> Result<(), ActionError> {
+    let osagactx = sagactx.user_data();
+    let params = sagactx.saga_params::<Params>()?;
+    let opctx = OpContext::for_saga_action(&sagactx, &params.serialized_authn);
+    osagactx
+        .datastore()
+        .project_delete_instance(&opctx, &params.authz_instance)
+        .await
+        .map_err(ActionError::action_failed)?;
+    Ok(())
+}
```

Comment:
It looks like part of what `project_delete_instance` is doing is going through an detaching disks. I guess that's really just updating the db record for the disks to unassociate them with the instance (via `Instance::detach_resource`). Looking in the detach disk call it's using the same `Instance::detach_resource`. I was wondering if that action, detaching disks, should be hoisted to the saga. Given that it's basically using the same thing under the hood I suppose not?

Reply:
I do not think it should be.

When I wrote `DatastoreDetachManyTarget::detach_resources` - which `Instance` objects implement, so they could detach disks - the whole point of it was that the detach operation (on disks) could be issued as part of the same statement that deletes instances, within a [common table expression](https://www.cockroachlabs.com/docs/stable/common-table-expressions.html) being issued to CRDB.

By existing within a single CTE, the statement is transactional, even though it modifies multiple tables and only sends a single request to CRDB. This follows the advice of https://www.cockroachlabs.com/docs/v22.2/performance-best-practices-overview#reduce-transaction-contention

Using CTEs, where possible, is generally our "best performance" way to combine multiple SQL statements. My usage of sagas elsewhere is a worse-performance attempt to get similar properties, in a non-atomic way.

Reply:
Related: Ideally, each function exposed from the `datastore()` would only issue a single DB statement, so that it either "fully completes" or "does nothing at all".

`project_delete_instance`, by virtue of being a CTE, already abides by this principle.